### PR TITLE
[DOC] Migrating Kafka Connect S2I to Kafka Connect

### PR DESCRIPTION
**Pull request by PaulRMellor**
Signed-off-by: prmellor <pmellor@redhat.com>

**Documentation**

New procedure for the _Using Guide_ to describe how to migrate from Kafka Connect S2I to Kafka Connect.
Also added a couple of notes, one in the _Deploying Guide_, to explain the forthcoming deprecation with a link to the new procedure.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+Now add configuration specific to the `KafkaConnect` resource to the new resource.
+
+. Add `build` configuration to configure all the connectors and other libraries you want to add to the Kafka Connect deployment.
++
+NOTE: Alternatively, you can build a new image with connectors manually, and specify it using the `.spec.image` property.
+
+. Delete the old `KafkaConnectS2I` resource:
++
+[source,shell,subs="+quotes"]
+kubectl delete _MY-KAFKA-CONNECT-S2I-CONFIG-FILE_
```
Comment (scholzj) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:50]: If we expect this to be a file, it should have `-f`:

 ```suggestion
kubectl delete -f _MY-KAFKA-CONNECT-S2I-CONFIG-FILE_
```

But the user might not have the file. So maybe we can add also `kubectl delete kafkaconnects2i _MY-KAFKA-CONNECT-S2I_`.

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
++
+[source,shell,subs="+quotes"]
+----
+kubectl get kctr --selector strimzi.io/cluster=_MY-CONNECT-CLUSTER_ -o name
+----
++
+Replace _MY-KAFKA-CONNECT-CLUSTER_ with the name of your Kafka Connect cluster.
++
+Connectors automatically recover through Kafka Connect storage.
+Even if you are using the Kafka Connect REST APi to manage them, you should not need to recreate them manually
```
Comment (scholzj) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:77]:
```suggestion
Even if you are using the Kafka Connect REST API to manage them, you should not need to recreate them manually.
```

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+. Deploy the new `KafkaConnect` resource:
++
+[source,shell,subs="+quotes"]
+kubectl apply -f _MY-KAFKA-CONNECT-CONFIG-FILE_
++
+Replace _MY-KAFKA-CONNECT-CONFIG-FILE_ with the name of the file containing your new `KafkaConnect` resource configuration.
++
+Wait until the new image is built, the deployment is created, and the pods have started.
+
+. Check that all expected connectors are present and are running:
```
Comment (scholzj) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:67]:
```suggestion
. If using KafkaConnector resources for managing Kafka Connect connectors, check that all expected connectors are present and are running:
```

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc (hunk 1 above)
Comment (PaulRMellor) [reply documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:50]: Adding both sounds sensible.

**Review by scholzj — APPROVED** [review]

**Review by ppatierno — APPROVED** [review]

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc [hunk 4]
```diff
@@ -0,0 +1,92 @@
+// Module included in the following assemblies:
+//
+// assembly-kafka-connect.adoc
+
+[id='proc-migrating-kafka-connect-s2i-{context}']
+= Migrating from Kafka Connect with S2I to Kafka Connect
+
+[role="_abstract"]
+Support for Kafka Connect with S2I and the `KafkaConnectS2I` resource is planned for deprecation.
+The decision follows the introduction of `build` configuration properties to the `KafkaConnect` resource,
```
Comment (laidan6000) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:10]: Not sure we should mention the decision here. How about "This follows..."

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+which are used to build a container image with the connector plugins you require for your data connections automatically.
+
+This procedure describes how to migrate your Kafka Connect with S2I instance to a standard Kafka Connect instance.
+To do this, you configure a new `KafkaConnect` custom resource to replace the `KafkaConnectS2I` resource, which is deleted.
```
Comment (laidan6000) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:14]: _Suggested change:_

Which is then deleted.

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+WARNING: The migration process involves downtime from the moment `KafkaConnectS2I` instance is deleted until the new `KafkaConnect` instance has been successfully deployed.
```
Comment (laidan6000) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:16]: _Suggested change:_

from the moment the `KafkaConnectS2I` instance...

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+.Prerequisites
+
+* Kafka Connect with S2I is deployed using a `KafkaConnectS2I` configuration
+* Kafka Connect with S2I is using an image with connectors added using an S2I build
+* Sink and source connector instances were created using `KafkaConnector` resources or the Kafka Connect REST API
+
+.Procedure
+
+. Create a new `KafkaConnect` custom resource using the same name as the name used for the `KafkaconnectS2I` resource.
+. Copy the `KafkaconnectS2I` resource properties to the `KafkaConnect` resource.
```
Comment (laidan6000) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:28]: Should they copy only the properties, or all of the YAML?

Comment (laidan6000) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:28]: Camel case: `KafkaConnectS2I`

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+. If specified, make sure you use the same `spec.config` properties:
++
+--
```
Comment (laidan6000) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:31]: Hmm, does these hyphens do anything?

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+* `group.id`
+* `offset.storage.topic`
+* `config.storage.topic`
+* `status.storage.topic`
++
+If these properties are not specified, defaults are used.
+In which case, leave them out of the `KafkaConnect` resource configuration as well.
+--
++
+Now add configuration specific to the `KafkaConnect` resource to the new resource.
```
Comment (laidan6000) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:41]: _Suggested change:_

Now add configuration specific to the `KafkaConnect` resource to the new `KafkaConnectS2I` resource.

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+Alternatively, you can specify the name of the resource:
++
+[source,shell,subs="+quotes"]
+kubectl delete kafkaconnects2i _MY-KAFKA-CONNECT-S2I_
++
+Replace _MY-KAFKA-CONNECT-S2I_ with the name of the `KafkaConnectS2I` resource.
++
+Wait until the Kafka Connect with S2I deployment and pods are deleted.
++
+IMPORTANT: No other resources must be deleted.
```
Comment (laidan6000) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:63]: _Suggested change:_

Use WARNING instead.

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+Wait until the new image is built, the deployment is created, and the pods have started.
+
+. If you are using `KafkaConnector` resources for managing Kafka Connect connectors, check that all expected connectors are present and are running:
++
+[source,shell,subs="+quotes"]
+----
+kubectl get kctr --selector strimzi.io/cluster=_MY-CONNECT-CLUSTER_ -o name
+----
++
+Replace _MY-KAFKA-CONNECT-CLUSTER_ with the name of your Kafka Connect cluster.
```
Comment (laidan6000) [inline documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:81]: _Suggested change:_

_MY-CONNECT-CLUSTER_

**Review by laidan6000 — COMMENTED** [review]
Nice procedure. I added some comments.

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc (hunk 7 above)
Comment (PaulRMellor) [reply documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:28]: It's pretty much just the kind that's different, so they could do either way and adjust accordingly.

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc (hunk 8 above)
Comment (PaulRMellor) [reply documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:31]: Keeps the formatting together so the final line in the step moves out where it needed to be

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc (hunk 9 above)
Comment (PaulRMellor) [reply documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:41]: it's not a new KafkaConnectS2I resource, it's a new KafkaConnect resource, so will leave as is

File: documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc (hunk 11 above)
Comment (PaulRMellor) [reply documentation/modules/configuring/proc-config-kafka-connect-s2i-migration.adoc:81]: We've used _My-KAFKA-CONNECT-..._ through the procedure, so will keep as is

**Comment by scholzj** [comment]
Sorry @laidan6000 ... looks like I merged it too early.

**Comment by laidan6000** [comment]
I thought something was weird because I couldn't suggest changes!

**Pull request merged**